README.md
# ![oneAPI](assets/oneapi-logo.png "oneAPI") Video Processing Library CPU Implementation

The oneAPI Video Processing Library (oneVPL) CPU Implementation is a runtime
implementation of oneVPL API for Intel CPUs. It provides video decode, encode,
and processing capabilities that run on the CPU.

See the [oneVPL Specification](https://spec.oneapi.io/versions/latest/elements/oneVPL/source/index.html) for additional information. This is part of the [oneAPI specification](https://www.oneapi.io/spec/).

---

**NOTE** 
- While the oneVPL GPU implementations are fully functional, the CPU implementation is currently for reference only and not intended to be competitive in terms of performance or features.
- This repository will only recieve critical bug fixes.  No updates to features, performance, etc. are planned. 
- Use of this implementation requires installation of the dispatcher provided as part of the [oneVPL base repository](https://github.com/oneapi-src/oneVPL).

---

## OneVPL Architecture
```mermaid
graph TD;
    VPL[oneVPL Dispatcher]-->oneVPL-cpu;
    VPL[oneVPL Dispatcher]-->oneVPL-intel-gpu;
    VPL[oneVPL Dispatcher]-->MediaSDK;
    VPL[oneVPL Dispatcher]-->Future1;
    VPL[oneVPL Dispatcher]-->Future2;
```

## CPU Implementation Feature/Color Format Info

This CPU implementation includes the following features:

- H.265/HEVC, H.264/AVC, MPEG2, and MJPEG decode (using libavcodec)
- AV1 decode (using libdav1d and libavcodec)
- MJPEG encode (using libavcodec)
- H.265/HEVC and AV1 encode (using Scalable Video Technology for HEVC, Scalable
  Video Technology for AV1, and libavcodec)
- H.264/AVC encode (using OpenH264/x264 and libavcodec)
- VPP - Crop, Resize, CSC (using libavfilter/ FFmpeg filters)

Note: H.265/HEVC, H.264, and AV1 are not available in 32-bit builds.

Codecs (+raw frame formats) supported by the CPU software implementation:

|  Codec        |  Encode        | Decode   |
|---------------|----------------|----------|
| MPEG2         |                |8 bit     |
|               |                |4:2:0 only|
| AVC/H.264     |8 bit (OpenH264)|          |
|               |8/10 bit (x264) |8/10 bit  |
|               |4:2:0 only      |4:2:0 only|
| HEVC/H.265    |8/10 bit        |8/10 bit  |
|               |4:2:0 only      |4:2:0 only|
| MJPEG         |8 bit           |8 bit     |
|               |4:2:0 only      |4:2:0 only|
| AV1           |8/10 bit        |8/10 bit  |
|               |4:2:0 only      |4:2:0 only|

Video processing (+raw frame formats) supported by the CPU software implementation:



| Operation     | In formats  | Out formats |
|---------------|-------------|-------------|
| Resize/Crop   | I420, I010  | I420, I010  |
|               | BGRA        | BGRA        |
| Colorspace    | I420, I010  | I420, I010  |
| Conversion    | BGRA        | BGRA        |

Note: I420 = 8 bit/420.  I010=10 bit/420.


## Installation
You can install oneVPL CPU implementation:

- from [oneVPL home page](https://software.intel.com/content/www/us/en/develop/tools/oneapi/components/onevpl.html) as a part of Intel&reg; oneAPI Base Toolkit or standalone.

### Installation from Source
See [Installation from Sources](INSTALL.md) for details.

## Usage

### Configure the Environment

If you install to standard system locations, applications can find the dispatcher library and
the dispatcher's default search rules will find your CPU implementation.

Otherwise you need to set up the environment search paths.  This is easiest to manage when the
install location <vpl-install-location> for oneVPL base is the same directory as used for the
CPU implementation.  In that case you can use the following steps:

For Linux:
```
source <vpl-install-location>/etc/vpl/vars.sh
```

For Windows:
```
<vpl-install-location>\etc\vpl\vars.bat
```

### Run the Command Line Tools

The oneVPL build that you installed as a prerequisite includes command line
tools that use installed implementations. You can use these command line tools
to process video from the command line using the CPU implementation.

All commands below assume `test/content` is the current directory.


- Report implementation capabilities:

    ```
    vpl-inspect
    ```

- Decode an an H.265 encoded video file:

    ```
    sample-decode h265 -i cars_320x240.h265 -o out.i420 -sw
    ```

- Encode a raw video file to H.265:

    ```
    sample-encode h265 -i cars_320x240.i420 -w 320 -h 240 -o out.h265
    ```


## Contributing

See [CONTRIBUTING.md](CONTRIBUTING.md) for more information.

## License

This project is licensed under the MIT License. See the [LICENSE](LICENSE) file
for details.

FFmpeg is an open source project licensed under LGPL and GPL. See [https://www.ffmpeg.org/legal.html](https://www.ffmpeg.org/legal.html). You are solely responsible for determining if your use of FFmpeg requires any additional licenses. Intel is not responsible for obtaining any such licenses, nor liable for any licensing fees due, in connection with your use of FFmpeg.

## Security

See the [Intel Security Center](https://www.intel.com/content/www/us/en/security-center/default.html) for information on how to report a potential
security issue or vulnerability.


CHANGELOG.md
# Changelog
All notable changes to the oneAPI Video Processing Library (oneVPL) CPU
Implementation will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [2023.2.0] - 2023-04-07

### Fixed

- Issues related to changes in open source artifact dependencies

## [2023.1.3] - 2023-03-03

### Fixed

- Issues detected from static analysis

## [2023.1.2] - 2023-01-27

### Changed

- Version of GoogleTest to 1.12.1

### Fixed

- Linker hardening flags being incorrectly set on Linux

## [2023.1.1] - 2022-12-23

### Changed

- README to add FFmpeg disclaimer.


## [2023.1.0] - 2022-11-10

### Fixed

- Wrong VPP out surface cropW/H info

### Added

- Changlog file

## [2023.0.0] - 2022-10-14

### Changed
- Versions of third party components used to more recent versions

### Fixed
- Build scripts not returning error on bad args

## [2022.2.0] - 2022-07-22

### Changed
- Build scripts to check for bash

## [2022.1.6] - 2022-07-08

### Added
- Additional code hardening on Windows*


## [2022.1.4] - 2022-05-27

### Changed

- NumFrameSuggested to increase by AsyncDepth

## [2022.1.3] - 2022-05-13

### Fixed

- Incorrect license being reported in capabilities query

## [2022.1.2] - 2022-04-29

### Changed

- Versions of third party components used to more recent versions


## [2022.1.1] - 2022-04-15

### Added

- Encode capabilities reporting support

## [2022.1.0] - 2022-03-18

### Changed

- `POSITION_INDEPENDENT_CODE` property to be set for all targets

## [2022.0.6] - 2022-03-04

### Changed

- Build to always support H264 encode by default

## [2022.0.5] - 2022-02-18

### Added
- Optional generation of IVF headers for AV1 encode

## [2022.0.4] - 2022-02-04

### Added
- H264 encoder with a permissive license

### Deprecated
- Support for Microsoft Visual Studio* 2017

## [2022.0.3] - 2022-01-21

### Added
- MPEG2 decode support

### Fixed
- HEVC encode crashing if GOP size = 1 with CPU implementation

## [2022.0.2] - 2021-12-16

### Added
- Extended device ID query support

## [2022.0.0] - 2021-12-06

### Changed
- Build scripts to be more robust and flexible

## [2021.6.0] - 2021-09-13

### Added
- oneVPL API 2.4 support
- HEVC 4:2:2 planar decode support

## [2021.5.0] - 2021-08-03

### Added
- Improved warnings for incompatible parameters

### Changed
- API version dependency to 2.4

### Fixed
- Numerous minor bugs

## [2021.4.0] - 2021-06-25

### Added

- oneVPL API 2.4 support

## [2021.2.2] - 2021-04-01

### Added
- oneVPL API 2.2 support
- 32-bit Windows* support
- Ubuntu 20.10 support

## [2021.1] - 2020-12-08

### Added
- CPU runtime implementation supporting oneVPL API 2.0
- H.264/AVC encode and decode support
- H.265/HEVC encode and decode support
- MJPEG encode and decode support
- AV1 encode and decode support


[Unreleased]: https://github.com/oneapi-src/oneVPL/compare/v2023.2.0...HEAD
[2023.2.0]: https://github.com/oneapi-src/oneVPL/compare/v2023.1.3...v2023.2.0
[2023.1.3]: https://github.com/oneapi-src/oneVPL-cpu/compare/v2023.1.2...v2023.1.3
[2023.1.2]: https://github.com/oneapi-src/oneVPL-cpu/compare/v2023.1.1...v2023.1.2
[2023.1.1]: https://github.com/oneapi-src/oneVPL-cpu/compare/v2023.1.0...v2023.1.1
[2023.1.0]: https://github.com/oneapi-src/oneVPL-cpu/compare/v2023.0.0...v2023.1.0
[2023.0.0]: https://github.com/oneapi-src/oneVPL-cpu/compare/v2022.2.4...v2023.0.0
[2022.2.4]: https://github.com/oneapi-src/oneVPL-cpu/compare/v2022.2.3...v2022.2.4
[2022.2.3]: https://github.com/oneapi-src/oneVPL-cpu/compare/v2022.2.2...v2022.2.3
[2022.2.2]: https://github.com/oneapi-src/oneVPL-cpu/compare/v2022.2.1...v2022.2.2
[2022.2.1]: https://github.com/oneapi-src/oneVPL-cpu/compare/v2022.2.0...v2022.2.1
[2022.2.0]: https://github.com/oneapi-src/oneVPL-cpu/compare/v2022.1.6...v2022.2.0
[2022.1.6]: https://github.com/oneapi-src/oneVPL-cpu/compare/v2022.1.5...v2022.1.6
[2022.1.5]: https://github.com/oneapi-src/oneVPL-cpu/compare/v2022.1.4...v2022.1.5
[2022.1.4]: https://github.com/oneapi-src/oneVPL-cpu/compare/v2022.1.3...v2022.1.4
[2022.1.3]: https://github.com/oneapi-src/oneVPL-cpu/compare/v2022.1.2...v2022.1.3
[2022.1.2]: https://github.com/oneapi-src/oneVPL-cpu/compare/v2022.1.1...v2022.1.2
[2022.1.1]: https://github.com/oneapi-src/oneVPL-cpu/compare/v2022.1.0...v2022.1.1
[2022.1.0]: https://github.com/oneapi-src/oneVPL-cpu/compare/v2022.0.6...v2022.1.0
[2022.0.6]: https://github.com/oneapi-src/oneVPL-cpu/compare/v2022.0.5...v2022.0.6
[2022.0.5]: https://github.com/oneapi-src/oneVPL-cpu/compare/v2022.0.4...v2022.0.5
[2022.0.4]: https://github.com/oneapi-src/oneVPL-cpu/compare/v2022.0.3...v2022.0.4
[2022.0.3]: https://github.com/oneapi-src/oneVPL-cpu/compare/v2022.0.2...v2022.0.3
[2022.0.2]: https://github.com/oneapi-src/oneVPL-cpu/compare/v2022.0.0...v2022.0.2
[2022.0.0]: https://github.com/oneapi-src/oneVPL-cpu/compare/v2021.6.0...v2022.0.0
[2021.6.0]: https://github.com/oneapi-src/oneVPL-cpu/compare/v2021.5.0...v2021.6.0
[2021.5.0]: https://github.com/oneapi-src/oneVPL-cpu/compare/v2021.4.0...v2021.5.0
[2021.4.0]: https://github.com/oneapi-src/oneVPL-cpu/compare/v2021.2.2...v2021.4.0
[2021.2.2]: https://github.com/oneapi-src/oneVPL-cpu/compare/v2021.1]:...v2021.2.2
[2021.1]: https://github.com/oneapi-src/oneVPL-cpu/releases/tag/v2021.1


api/vpl/mfxdefs.h
/*############################################################################
  # Copyright (C) 2019-2020 Intel Corporation
  #
  # SPDX-License-Identifier: MIT
  ############################################################################*/

#ifndef __MFXDEFS_H__
#define __MFXDEFS_H__

#define MFX_VERSION_MAJOR 2
#define MFX_VERSION_MINOR 0

// MFX_VERSION_NEXT is always +1 from last public release
// may be enforced by MFX_VERSION_USE_LATEST define
// if MFX_VERSION_USE_LATEST is defined  MFX_VERSION is ignored

#define MFX_VERSION_NEXT (MFX_VERSION_MAJOR * 1000 + MFX_VERSION_MINOR + 1)

// MFX_VERSION - version of API that 'assumed' by build may be provided externally
// if it omitted then latest stable API derived from Major.Minor is assumed


#if !defined(MFX_VERSION)
  #if defined(MFX_VERSION_USE_LATEST)
    #define MFX_VERSION MFX_VERSION_NEXT
  #else
    #define MFX_VERSION (MFX_VERSION_MAJOR * 1000 + MFX_VERSION_MINOR)
  #endif
#else
  #undef MFX_VERSION_MINOR
  #define MFX_VERSION_MINOR ((MFX_VERSION) % 1000)
#endif

/*! This is correspondent version of MediaSDK Legacy API which is used as a basis 
   for the current oneVPL API. */

#define MFX_LEGACY_VERSION 1034


#ifdef __cplusplus
extern "C"
{
#endif /* __cplusplus */

/* In preprocessor syntax # symbol has stringize meaning,
   so to expand some macro to preprocessor pragma we need to use
   special compiler dependent construction */

#if defined(_MSC_VER)
    #define MFX_PRAGMA_IMPL(x) __pragma(x)
#else
    #define MFX_PRAGMA_IMPL(x) _Pragma(#x)
#endif

#define MFX_PACK_BEGIN_X(x) MFX_PRAGMA_IMPL(pack(push, x))
#define MFX_PACK_END()      MFX_PRAGMA_IMPL(pack(pop))

/* The general rule for alignment is following:
   - structures with pointers have 4/8 bytes alignment on 32/64 bit systems
   - structures with fields of type mfxU64/mfxF64 (unsigned long long / double)
     have alignment 8 bytes on 64 bit and 32 bit Windows, on Linux alignment is 4 bytes
   - all the rest structures are 4 bytes aligned
   - there are several exceptions: some structs which had 4-byte alignment were extended
     with pointer / long type fields; such structs have 4-byte alignment to keep binary
     compatibility with previously release API */

#define MFX_PACK_BEGIN_USUAL_STRUCT()        MFX_PACK_BEGIN_X(4)

/* 64-bit LP64 data model */
#if defined(_WIN64) || defined(__LP64__)
    #define MFX_PACK_BEGIN_STRUCT_W_PTR()    MFX_PACK_BEGIN_X(8)
    #define MFX_PACK_BEGIN_STRUCT_W_L_TYPE() MFX_PACK_BEGIN_X(8)
/* 32-bit ILP32 data model Windows (Intel architecture) */
#elif defined(_WIN32) || defined(_M_IX86) && !defined(__linux__)
    #define MFX_PACK_BEGIN_STRUCT_W_PTR()    MFX_PACK_BEGIN_X(4)
    #define MFX_PACK_BEGIN_STRUCT_W_L_TYPE() MFX_PACK_BEGIN_X(8)
/* 32-bit ILP32 data model Linux */
#elif defined(__ILP32__)
    #define MFX_PACK_BEGIN_STRUCT_W_PTR()    MFX_PACK_BEGIN_X(4)
    #define MFX_PACK_BEGIN_STRUCT_W_L_TYPE() MFX_PACK_BEGIN_X(4)
#else
    #error Unknown packing
#endif

  #define __INT64   long long
  #define __UINT64  unsigned long long

#ifdef _WIN32
  #define MFX_CDECL __cdecl
  #define MFX_STDCALL __stdcall
#else
  #define MFX_CDECL
  #define MFX_STDCALL
#endif /* _WIN32 */

#define MFX_INFINITE 0xFFFFFFFF

typedef unsigned char       mfxU8;         /*!< Unsigned integer, 8 bit type. */
typedef char                mfxI8;         /*!< Signed integer, 8 bit type. */
typedef short               mfxI16;        /*!< Signed integer, 16 bit type. */
typedef unsigned short      mfxU16;        /*!< Unsigned integer, 16 bit type. */
typedef unsigned int        mfxU32;        /*!< Unsigned integer, 32 bit type. */
typedef int                 mfxI32;        /*!< Signed integer, 32 bit type. */
#if defined( _WIN32 ) || defined ( _WIN64 )
typedef unsigned long       mfxUL32;       /*!< Unsigned integer, 32 bit type. */
typedef long                mfxL32;        /*!< Signed integer, 32 bit type. */
#else
typedef unsigned int        mfxUL32;       /*!< Unsigned integer, 32 bit type. */
typedef int                 mfxL32;        /*!< Signed integer, 32 bit type. */
#endif
typedef float               mfxF32;        /*!< Single-precision floating point, 32 bit type. */
typedef double              mfxF64;        /*!< Double-precision floating point, 64 bit type. */
typedef __UINT64            mfxU64;        /*!< Unsigned integer, 64 bit type. */
typedef __INT64             mfxI64;        /*!< Signed integer, 64 bit type. */
typedef void*               mfxHDL;        /*!< Handle type. */
typedef mfxHDL              mfxMemId;      /*!< Memory ID type. */
typedef void*               mfxThreadTask; /*!< Thread task type. */
typedef char                mfxChar;       /*!< UTF-8 byte. */

/* MFX structures version info */
MFX_PACK_BEGIN_USUAL_STRUCT()
/*! Introduce field Version for any structures.
Minor number is incremented when reserved fields are used,
major number is incremented when size of structure is increased.
Assumed that any structure changes are backward binary compatible.
 mfxStructVersion starts from {1,0} for any new API structures, if mfxStructVersion is
 added to the existent legacy structure (replacing reserved fields) it starts from {1, 1}.
*/
typedef union {
    /*! Structure with Major and Minor fields.  */
    /*! @struct Anonymous */
    struct {
        mfxU8  Minor; /*!< Minor number of the correspondent structure. */
        mfxU8  Major; /*!< Major number of the correspondent structure. */
    };
    mfxU16  Version;   /*!< Structure version number */
} mfxStructVersion;
MFX_PACK_END()

#define MFX_STRUCT_VERSION(MAJOR, MINOR) (256*(MAJOR) + (MINOR))


#define MFX_VARIANT_VERSION MFX_STRUCT_VERSION(1, 0)

/*! The mfxVariantType enumerator data types for mfxVarianf type. */
typedef enum {
    MFX_VARIANT_TYPE_UNSET = 0, /*!< Undefined type. */
    MFX_VARIANT_TYPE_U8 = 1,   /*!< 8-bit unsigned integer. */
    MFX_VARIANT_TYPE_I8,       /*!< 8-bit signed integer. */
    MFX_VARIANT_TYPE_U16,      /*!< 16-bit unsigned integer. */
    MFX_VARIANT_TYPE_I16,      /*!< 16-bit signed integer. */
    MFX_VARIANT_TYPE_U32,      /*!< 32-bit unsigned integer. */
    MFX_VARIANT_TYPE_I32,      /*!< 32-bit signed integer. */
    MFX_VARIANT_TYPE_U64,      /*!< 64-bit unsigned integer. */
    MFX_VARIANT_TYPE_I64,      /*!< 64-bit signed integer. */
    MFX_VARIANT_TYPE_F32,      /*!< 32-bit single precision floating point. */
    MFX_VARIANT_TYPE_F64,      /*!< 64-bit double precision floating point. */
    MFX_VARIANT_TYPE_PTR,      /*!< Generic type pointer. */
} mfxVariantType;

MFX_PACK_BEGIN_STRUCT_W_PTR()
/*! The mfxVariantType enumerator data types for mfxVarianf type. */
typedef struct {
    mfxStructVersion Version;    /*!< Version of the structure. */
    mfxVariantType   Type;       /*!< Value type. */
    /*! Value data holder. */
    union data {
        mfxU8   U8; /*!< mfxU8 data. */
        mfxI8   I8; /*!< mfxI8 data. */
        mfxU16 U16; /*!< mfxU16 data. */
        mfxI16 I16; /*!< mfxI16 data. */
        mfxU32 U32; /*!< mfxU32 data. */
        mfxI32 I32; /*!< mfxI32 data. */
        mfxU64 U64; /*!< mfxU64 data. */
        mfxI64 I64; /*!< mfxI64 data. */
        mfxF32 F32; /*!< mfxF32 data. */
        mfxF64 F64; /*!< mfxF64 data. */
       mfxHDL Ptr;  /*!< Pointer. */
    } Data;         /*!< Value data member. */
} mfxVariant;
MFX_PACK_END()

MFX_PACK_BEGIN_USUAL_STRUCT()
/*! This structure represents a range of unsigned values */
typedef struct {
    mfxU32 Min;  /*!< Minimal value of the range. */
    mfxU32 Max;  /*!< Maximal value of the range. */
    mfxU32 Step; /*!< Value incrementation step. */
} mfxRange32U;
MFX_PACK_END()


/*! Thus structure represents a pair of numbers of type mfxI16. */
typedef struct {
    mfxI16  x; /*!< First number. */
    mfxI16  y; /*!< Second number. */
} mfxI16Pair;

/*! Thus structure represents pair of handles of type mfxHDL. */
typedef struct {
    mfxHDL first;  /*!< First handle. */
    mfxHDL second; /*!< Second handle. */
} mfxHDLPair;


/*********************************************************************************\
Error message
\*********************************************************************************/
/*! @enum mfxStatus Itemizes status codes returned by SDK functions. */
typedef enum
{
    /* no error */
    MFX_ERR_NONE                        = 0,    /*!< No error. */
    /* reserved for unexpected errors */
    MFX_ERR_UNKNOWN                     = -1,   /*!< Unknown error. */

    /* error codes <0 */
    MFX_ERR_NULL_PTR                    = -2,   /*!< Null pointer. */
    MFX_ERR_UNSUPPORTED                 = -3,   /*!< Unsupported feature. */
    MFX_ERR_MEMORY_ALLOC                = -4,   /*!< Failed to allocate memory. */
    MFX_ERR_NOT_ENOUGH_BUFFER           = -5,   /*!< Insufficient buffer at input/output. */
    MFX_ERR_INVALID_HANDLE              = -6,   /*!< Invalid handle. */
    MFX_ERR_LOCK_MEMORY                 = -7,   /*!< Failed to lock the memory block. */
    MFX_ERR_NOT_INITIALIZED             = -8,   /*!< Member function called before initialization. */
    MFX_ERR_NOT_FOUND                   = -9,   /*!< The specified object is not found. */
    MFX_ERR_MORE_DATA                   = -10,  /*!< Expect more data at input. */
    MFX_ERR_MORE_SURFACE                = -11,  /*!< Expect more surface at output. */
    MFX_ERR_ABORTED                     = -12,  /*!< Operation aborted. */
    MFX_ERR_DEVICE_LOST                 = -13,  /*!< Lose the hardware acceleration device. */
    MFX_ERR_INCOMPATIBLE_VIDEO_PARAM    = -14,  /*!< Incompatible video parameters. */
    MFX_ERR_INVALID_VIDEO_PARAM         = -15,  /*!< Invalid video parameters. */
    MFX_ERR_UNDEFINED_BEHAVIOR          = -16,  /*!< Undefined behavior. */
    MFX_ERR_DEVICE_FAILED               = -17,  /*!< Device operation failure. */
    MFX_ERR_MORE_BITSTREAM              = -18,  /*!< Expect more bitstream buffers at output. */
    MFX_ERR_GPU_HANG                    = -21,  /*!< Device operation failure caused by GPU hang. */
    MFX_ERR_REALLOC_SURFACE             = -22,  /*!< Bigger output surface required. */
    MFX_ERR_RESOURCE_MAPPED             = -23,  /*!< Write access is already acquired and user requested
                                                   another write access, or read access with MFX_MEMORY_NO_WAIT flag. */
    MFX_ERR_NOT_IMPLEMENTED             = -24,   /*!< Feature or function not implemented. */
    /* warnings >0 */
    MFX_WRN_IN_EXECUTION                = 1,    /*!< The previous asynchronous operation is in execution. */
    MFX_WRN_DEVICE_BUSY                 = 2,    /*!< The hardware acceleration device is busy. */
    MFX_WRN_VIDEO_PARAM_CHANGED         = 3,    /*!< The video parameters are changed during decoding. */
    MFX_WRN_PARTIAL_ACCELERATION        = 4,    /*!< Software acceleration is used. */
    MFX_WRN_INCOMPATIBLE_VIDEO_PARAM    = 5,    /*!< Incompatible video parameters. */
    MFX_WRN_VALUE_NOT_CHANGED           = 6,    /*!< The value is saturated based on its valid range. */
    MFX_WRN_OUT_OF_RANGE                = 7,    /*!< The value is out of valid range. */
    MFX_WRN_FILTER_SKIPPED              = 10,   /*!< One of requested filters has been skipped. */
    /* low-delay partial output */
    MFX_ERR_NONE_PARTIAL_OUTPUT         = 12,   /*!< Frame is not ready, but bitstream contains partial output. */

    /* threading statuses */
    MFX_TASK_DONE = MFX_ERR_NONE,               /*!< Task has been completed. */
    MFX_TASK_WORKING                    = 8,    /*!< There is some more work to do. */
    MFX_TASK_BUSY                       = 9,    /*!< Task is waiting for resources. */

    /* plug-in statuses */
    MFX_ERR_MORE_DATA_SUBMIT_TASK       = -10000, /*!< Return MFX_ERR_MORE_DATA but submit internal asynchronous task. */

} mfxStatus;


// Application 
#if defined(MFX_DISPATCHER_EXPOSED_PREFIX)

#include "mfxdispatcherprefixedfunctions.h"

#endif // MFX_DISPATCHER_EXPOSED_PREFIX


#ifdef __cplusplus
}
#endif /* __cplusplus */

#endif /* __MFXDEFS_H__ */
